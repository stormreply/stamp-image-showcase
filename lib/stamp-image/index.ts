import { StampImageBucket } from './bucket';
import { StampImageLamnbda } from './lambda';

export { StampImageBucket, StampImageLamnbda };
