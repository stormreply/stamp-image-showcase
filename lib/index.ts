import * as cdk from '@aws-cdk/core';

import { StampImageLamnbda, StampImageBucket } from './stamp-image';

export class StampImageStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // 1. Create lambda function for stamping images
        const handler = new StampImageLamnbda(this, 'Function', {
            functionName: 'StampImageFunction',
            handler: 'app.handler',
            timeout: cdk.Duration.seconds(60),
            memorySize: 1024,
        });

        // 2. Create s3 bucket as image source and target
        new StampImageBucket(this, 'Bucket', {
            handler,
        });
    }
}
