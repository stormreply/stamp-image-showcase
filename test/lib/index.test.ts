import { SynthUtils } from '@aws-cdk/assert';

import * as cdk from '@aws-cdk/core';

import { StampImageStack } from '../../lib';

test('Stack Snapshot', () => {
    const app = new cdk.App();

    // WHEN

    const stack = new StampImageStack(app, 'MyTestStack');

    // THEN

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
