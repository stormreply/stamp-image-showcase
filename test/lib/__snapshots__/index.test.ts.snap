// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dArtifactHash8CF4E94F": Object {
      "Description": "Artifact hash for asset \\"f57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92d\\"",
      "Type": "String",
    },
    "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dS3Bucket2E71D38A": Object {
      "Description": "S3 bucket for asset \\"f57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92d\\"",
      "Type": "String",
    },
    "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dS3VersionKey7523E01A": Object {
      "Description": "S3 key for asset version \\"f57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92d\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketNotifications8F2E257D": Object {
      "DependsOn": Array [
        "FunctionAllowBucketNotificationsFromMyTestStackBucketA865EC6BDC2C107D",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "Bucket83908E77",
        },
        "NotificationConfiguration": Object {
          "LambdaFunctionConfigurations": Array [
            Object {
              "Events": Array [
                "s3:ObjectCreated:*",
              ],
              "Filter": Object {
                "Key": Object {
                  "FilterRules": Array [
                    Object {
                      "Name": "prefix",
                      "Value": "input/",
                    },
                  ],
                },
              },
              "LambdaFunctionArn": Object {
                "Fn::GetAtt": Array [
                  "Function76856677",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": Object {
      "DependsOn": Array [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = (event, context) => {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, import/no-extraneous-dependencies
    const s3 = new (require('aws-sdk').S3)();
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const https = require('https');
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const url = require('url');
    log(JSON.stringify(event, undefined, 2));
    const props = event.ResourceProperties;
    if (event.RequestType === 'Delete') {
        props.NotificationConfiguration = {}; // this is how you clean out notifications
    }
    const req = {
        Bucket: props.BucketName,
        NotificationConfiguration: props.NotificationConfiguration,
    };
    return s3.putBucketNotificationConfiguration(req, (err, data) => {
        log({ err, data });
        if (err) {
            return submitResponse('FAILED', err.message + \`\\\\nMore information in CloudWatch Log Stream: \${context.logStreamName}\`);
        }
        else {
            return submitResponse('SUCCESS');
        }
    });
    function log(obj) {
        console.error(event.RequestId, event.StackId, event.LogicalResourceId, obj);
    }
    // tslint:disable-next-line:max-line-length
    // adapted from https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html#cfn-lambda-function-code-cfnresponsemodule
    // to allow sending an error messge as a reason.
    function submitResponse(responseStatus, reason) {
        const responseBody = JSON.stringify({
            Status: responseStatus,
            Reason: reason || 'See the details in CloudWatch Log Stream: ' + context.logStreamName,
            PhysicalResourceId: event.PhysicalResourceId || event.LogicalResourceId,
            StackId: event.StackId,
            RequestId: event.RequestId,
            LogicalResourceId: event.LogicalResourceId,
            NoEcho: false,
        });
        log({ responseBody });
        const parsedUrl = url.parse(event.ResponseURL);
        const options = {
            hostname: parsedUrl.hostname,
            port: 443,
            path: parsedUrl.path,
            method: 'PUT',
            headers: {
                'content-type': '',
                'content-length': responseBody.length,
            },
        };
        const request = https.request(options, (r) => {
            log({ statusCode: r.statusCode, statusMessage: r.statusMessage });
            context.done();
        });
        request.on('error', (error) => {
            log({ sendError: error });
            context.done();
        });
        request.write(responseBody);
        request.end();
    }
};",
        },
        "Description": "AWS CloudFormation handler for \\"Custom::S3BucketNotifications\\" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": Array [
          Object {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Function76856677": Object {
      "DependsOn": Array [
        "FunctionServiceRoleDefaultPolicy2F49994A",
        "FunctionServiceRole675BB04A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dS3Bucket2E71D38A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dS3VersionKey7523E01A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf57dc25a3a373da576a0b01a6210243dea11887f5e1a325c51fabe479458e92dS3VersionKey7523E01A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "app.handler",
        "Layers": Array [
          Object {
            "Fn::GetAtt": Array [
              "FunctionApplication0DE0BB6D",
              "Outputs.LayerVersion",
            ],
          },
        ],
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "FunctionServiceRole675BB04A",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionAllowBucketNotificationsFromMyTestStackBucketA865EC6BDC2C107D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Function76856677",
            "Arn",
          ],
        },
        "Principal": "s3.amazonaws.com",
        "SourceAccount": Object {
          "Ref": "AWS::AccountId",
        },
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "Bucket83908E77",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "FunctionApplication0DE0BB6D": Object {
      "Properties": Object {
        "Location": Object {
          "ApplicationId": "arn:aws:serverlessrepo:us-east-1:145266761615:applications/image-magick-lambda-layer",
          "SemanticVersion": "1.0.0",
        },
      },
      "Type": "AWS::Serverless::Application",
    },
    "FunctionServiceRole675BB04A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FunctionServiceRoleDefaultPolicy2F49994A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "Bucket83908E77",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "Bucket83908E77",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FunctionServiceRoleDefaultPolicy2F49994A",
        "Roles": Array [
          Object {
            "Ref": "FunctionServiceRole675BB04A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Transform": "AWS::Serverless-2016-10-31",
}
`;
